streams/js/src/main/scala/zio/stream/platform.scala

```diff
@@ -2,12 +2,132 @@ package zio.stream
+                              runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer).unit)
+                            } catch {
+                              case FiberFailure(c) if c.interrupted =>
+                                Future.successful(())
+                            }
+                          )
+                        }
+        pull <- eitherStream match {
+                  case Left(canceler) =>
+                    (for {
+                      done <- ZRef.makeManaged(false)
+                    } yield done.get.flatMap {
+                      if (_) Pull.end
+                      else
+                        output.take.flatMap(_.done).onError(_ => done.set(true) *> output.shutdown)
+                    }).ensuring(canceler)
+                  case Right(stream) => output.shutdown.toManaged_ *> stream.process
+                }
+      } yield pull
+    }
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times
+   * The registration of the callback itself returns an effect. The optionality of the
+   * error type `E` can be used to signal the end of the stream, by setting it to `None`.
+   */
+  def effectAsyncM[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Unit]) => ZIO[R, E, Any],
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    managed {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        _ <- register { k =>
+               try {
+                 runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer).unit)
+               } catch {
+                 case FiberFailure(c) if c.interrupted =>
+                   Future.successful(())
+               }
+             }.toManaged_
+        done <- ZRef.makeManaged(false)
+        pull = done.get.flatMap {
+                 if (_)
+                   Pull.end
+                 else
+                   output.take.flatMap(_.done).onError(_ => done.set(true) *> output.shutdown)
+               }
+      } yield pull
+    }.flatMap(repeatEffectChunkOption(_))
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The registration of the callback can possibly return the stream synchronously.
+   * The optionality of the error type `E` can be used to signal the end of the stream,
+   * by setting it to `None`.
+   */
+  def effectAsyncMaybe[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => Option[ZStream[R, E, A]],
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    ZStream {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        maybeStream <- ZManaged.effectTotal {
+                         register(k => runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer)))
```

Comment:
Do we want to handle interruption from `output.offer` here too?

Reply:
Good catch, updated.